plugins/package-managers/bazel/src/main/kotlin/Bazel.kt

```diff
@@ -0,0 +1,221 @@
+        override val globsForDefinitionFiles = listOf("MODULE", "MODULE.bazel")
+
+        override fun create(
+            analysisRoot: File,
+            analyzerConfig: AnalyzerConfiguration,
+            repoConfig: RepositoryConfiguration
+        ) = Bazel(type, analysisRoot, analyzerConfig, repoConfig)
+    }
+
+    override fun command(workingDir: File?) = "bazel"
+
+    override fun getVersionArguments() = "version"
+
+    override fun transformVersion(output: String) = transformBazelVersion(output)
+
+    // Bazel 6.0 already supports bzlmod but it is not enabled by default.
+    // Supporting it would require adding the flag "--enable_bzlmod=true" at the correct position of all bazel
+    // invocations.
+    override fun getVersionRequirement(): RangesList = RangesListFactory.create(">=7.0")
+
+    override fun resolveDependencies(definitionFile: File, labels: Map<String, String>): List<ProjectAnalyzerResult> {
+        val projectDir = definitionFile.parentFile
+        val lockFile = projectDir.resolve(LOCKFILE_NAME)
+        val projectVcs = processProjectVcs(projectDir)
+        val moduleMetadata = Parser(definitionFile.readText()).parse()
+        val depDirectives = moduleMetadata.dependencies.associateBy { "${it.name}@${it.version}" }
+        val scopes = getDependencyGraph(projectDir, depDirectives)
+
+        // If no lock file is present, getDependencyGraph() runs "bazel mod graph", which creates a MODULE.bazel.lock
+        // file as a side effect. That file contains the URL of the Bazel module registry that was used for dependency
+        // resolution.
+        val lockFileData = parseLockfile(lockFile)
+
+        val bmrClient = BazelModuleRegistryClient.create(lockFileData.registryUrl())
+        val packages = scopes.flatMap { getPackages(it.dependencies, bmrClient) }.filterNotNull()
+
+        return listOf(
+            ProjectAnalyzerResult(
+                project = Project(
+                    id = Identifier(
+                        type = managerName,
+                        name = moduleMetadata.module?.name ?: "root",
+                        version = moduleMetadata.module?.version.orEmpty(),
+                        namespace = ""
+                    ),
+                    definitionFilePath = VersionControlSystem.getPathInfo(definitionFile).path,
+                    authors = emptySet(),
+                    declaredLicenses = emptySet(),
+                    vcs = projectVcs,
+                    vcsProcessed = projectVcs,
+                    homepageUrl = "",
+                    scopeDependencies = scopes
+                ),
+                packages = packages.toSet()
+            )
+        )
+    }
+
+    private fun getPackages(
+        graph: Set<PackageReference>,
+        bmrClient: BazelModuleRegistryClient,
+        seen: MutableSet<Identifier> = mutableSetOf()
+    ): Set<Package?> {
+        val result = mutableSetOf<Package?>()
```

Comment:
You may utilize `buildSet` and turn this function into an extension function, e.g. 
`getPackages(..): Set<Package> = buildSet { ... }` 

This would separate concerns: 
1. Obtaining a set of identifiers from given package references
2. Creating a package for a given id

Reply:
Actually I believe it'd be nice to additionally separate also the retrieval of the `ModuleMetadata`, to arrive at something like:
```
val ids = packageRefs.....// some mapping code from the refs to the ids
val moduleMetadataForId = ids.map { getModuleMetedata(it) }
val packages = ids.map { id -> getPackage(id, moduleMetadataForId[id]) }
```

This would make it easier to parallelize the querying later on, because no shared mutable data structure is involved anymore. After writing noticed that moduleSourceInfo retrieval is missing and would need to be incorporated above in analog way.

Reply:
Not directly related, but I just realized that there are now two data classes called `ModuleMetadata`. One in `StarklarkParser` and one in `BazelModuleRegistryClient`. Seems quite confusing. Do you have any thoughts on this?

Reply:
> Actually I believe it'd be nice to additionally separate also the retrieval of the `ModuleMetadata`, to arrive at something like:
> 
> ```
> val ids = packageRefs.....// some mapping code from the refs to the ids
> val moduleMetadataForId = ids.map { getModuleMetedata(it) }
> val packages = ids.map { id -> getPackage(id, moduleMetadataForId[id]) }
> ```

I've made some changes that I believe implement what you suggested. I didn't use `buildSet` though.

I guess failure to retrieve data from the module registry should result in `Issue`s being created. However, I don't know how to implement that without using a shared mutable set that is either passed around or an instance variable.

Reply:
I haven't considered the issue before. As it's not implemented, shall we leave this consideration aside for this iteration?

Reply:
> I haven't considered the issue before. As it's not implemented, shall we leave this consideration aside for this iteration?

Works for me.